api/core/v1alpha1/etcdopstask.go

```diff
@@ -0,0 +1,206 @@
+
+	// EtcdName refers to the name of the Etcd resource that this task will operate on.
+	// +optional
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="etcdName is immutable"
+	EtcdName *string `json:"etcdName"`
+}
+
+// EtcdOpsTaskConfig holds the configuration for the specific operation.
+// Exactly one of its members must be set according to the operation to be performed.
+// +kubebuilder:validation:MinProperties=1
+// +kubebuilder:validation:MaxProperties=1
+type EtcdOpsTaskConfig struct {
+	// OnDemandSnapshot defines the configuration for an on-demand snapshot task.
+	// +optional
+	OnDemandSnapshot *OnDemandSnapshotConfig `json:"onDemandSnapshot,omitempty"`
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Status
+////////////////////////////////////////////////////////////////////////////////
+
+// EtcdOpsTaskStatus defines the observed state of an EtcdOpsTask.
+type EtcdOpsTaskStatus struct {
+	// State is the overall state of the task.
+	// The controller initializes this field when processing the task.
+	// +optional
+	State *TaskState `json:"state,omitempty"`
+
+	// LastTransitionTime is the last time the state transitioned from one value to another.
+	// +optional
+	LastTransitionTime *metav1.Time `json:"lastTransitionTime,omitempty"`
+
+	// StartedAt is the time at which the task transitioned from Pending to InProgress.
+	// +optional
+	StartedAt *metav1.Time `json:"startedAt,omitempty"`
+
+	// Phase represents the current phase of the task's lifecycle.
+	// +optional
+	Phase *OperationPhase `json:"phase,omitempty"`
+
+	// LastErrors is a list of the most recent errors observed during the task's execution.
+	// A maximum of 10 latest errors will be recorded.
+	// +optional
+	// +kubebuilder:validation:MaxItems=10
+	LastErrors []LastError `json:"lastErrors,omitempty"`
+
+	// LastOperation tracks the fine-grained progress of the task's execution.
+	// The controller initializes this field when processing the task.
+	// +optional
+	LastOperation *LastOperation `json:"lastOperation,omitempty"`
+}
+
+// GetEtcdReference returns the NamespacedName of the etcd object referenced by the task.
+func (t *EtcdOpsTask) GetEtcdReference() (types.NamespacedName, error) {
+	if t.Spec.EtcdName == nil || *t.Spec.EtcdName == "" {
+		return types.NamespacedName{}, fmt.Errorf("etcdName is required but not specified")
+	}
+	return types.NamespacedName{
+		Name:      *t.Spec.EtcdName,
+		Namespace: t.Namespace,
+	}, nil
+}
+
+// IsCompleted returns true if the task is completed.
+func (t *EtcdOpsTask) IsCompleted() bool {
+	if t.Status.State == nil {
+		return false
+	}
+	return *t.Status.State == TaskStateSucceeded || *t.Status.State == TaskStateFailed || *t.Status.State == TaskStateRejected
+}
+
+// IsMarkedForDeletion returns true if the deletion timestamp is set.
+func (t *EtcdOpsTask) IsMarkedForDeletion() bool {
```

Comment:
Code duplication. We already have a helper in `core/v1alpha1/helper.go`
```go
func IsEtcdMarkedForDeletion(etcdObjMeta metav1.ObjectMeta) bool {
	return etcdObjMeta.DeletionTimestamp != nil
}
```
You can make it generic:
```go
func IsResourceMarkedForDeletion(objMeta metav1.ObjectMeta) bool {
	return objMeta.DeletionTimestamp != nil
}
```
Use `IsResourceMarkedForDeletion` in all places.